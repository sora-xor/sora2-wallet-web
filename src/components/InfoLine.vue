<template>
  <div class="info-line">
    <slot name="info-line-prefix" />
    <span class="info-line-label">{{ label }}</span>
    <s-tooltip
      v-if="labelTooltip"
      popper-class="info-tooltip info-tooltip--info-line"
      :content="labelTooltip"
      placement="right-start"
      border-radius="mini"
    >
      <s-icon name="info-16" size="14px" />
    </s-tooltip>
    <div class="info-line-content">
      <template v-if="isValueExists">
        <slot name="info-line-value-prefix" />
        <formatted-amount
          v-if="isFormatted"
          class="info-line-value"
          :value="value"
          :asset-symbol="assetSymbol"
          :font-size-rate="formattedFontSize"
          :font-weight-rate="formattedFontWeight"
          :value-can-be-hidden="valueCanBeHidden"
        />
        <span v-else-if="!valueCanBeHidden || !shouldBalanceBeHidden" class="info-line-value">
          {{ value }}
          <span v-if="assetSymbol" class="asset-symbol">{{ ' ' + assetSymbol }}</span>
        </span>
        <span v-else class="info-line-value">{{ HiddenValue }}</span>
        <formatted-amount
          v-if="fiatValue"
          is-fiat-value
          with-left-shift
          :value="fiatValue"
          :font-size-rate="formattedFontSize"
          :value-can-be-hidden="valueCanBeHidden"
        />
      </template>
      <slot />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'vue-property-decorator';
import { Getter } from 'vuex-class';

import FormattedAmount from './FormattedAmount.vue';
import { FontSizeRate, FontWeightRate, HiddenValue } from '../consts';

@Component({
  components: { FormattedAmount },
})
export default class InfoLine extends Vue {
  readonly HiddenValue = HiddenValue;

  @Prop({ default: '', type: String }) readonly label!: string;
  @Prop({ default: '', type: String }) readonly labelTooltip!: string;
  @Prop({ default: '' }) readonly value!: string;
  @Prop({ default: '', type: String }) readonly assetSymbol!: string;
  @Prop({ default: false, type: Boolean }) readonly isFormatted!: boolean;
  @Prop({ default: '', type: String }) readonly fiatValue!: string;
  /**
   * Define directly that this field displays value which can be hidden by hide balances button.
   */
  @Prop({ default: false, type: Boolean }) readonly valueCanBeHidden!: boolean;

  @Getter shouldBalanceBeHidden!: boolean;

  get isValueExists(): boolean {
    if (this.value === 'NaN' || this.value.includes('Infinity')) {
      console.warn(`The ${this.label} value is: ${this.value}.`);
      return false;
    }
    return !!this.value;
  }

  get formattedFontSize(): Nullable<FontSizeRate> {
    return this.isFormatted ? FontSizeRate.MEDIUM : null;
  }

  get formattedFontWeight(): Nullable<FontWeightRate> {
    return this.isFormatted ? FontWeightRate.SMALL : null;
  }
}
</script>

<style lang="scss">
.info-tooltip--info-line {
  margin-top: -10px;
  .popper__arrow {
    margin-top: #{$basic-spacing-mini};
  }
}
.info-line {
  &-container {
    border-radius: var(--s-border-radius-small);
    margin-top: var(--s-basic-spacing);
    padding: var(--s-basic-spacing) 0 0;
    width: 100%;

    &__title {
      font-size: var(--s-heading6-font-size);
      font-weight: 300;
      line-height: var(--s-line-height-medium);
      letter-spacing: var(--s-letter-spacing-small);
      color: var(--s-color-base-content-secondary);
      text-transform: uppercase;
      margin-bottom: #{$basic-spacing-small};
    }
  }
}
</style>

<style lang="scss" scoped>
.info-line {
  display: flex;
  align-items: center;
  width: 100%;
  padding: #{$basic-spacing-extra-mini} #{$basic-spacing-mini};
  color: var(--s-color-base-content-primary);
  font-size: var(--s-font-size-extra-small);
  line-height: var(--s-line-height-small);
  border-bottom: 1px solid var(--s-color-base-border-secondary);
  & + .info-line {
    margin-top: #{$basic-spacing-small};
  }
  &:first-child {
    margin-top: 0;
  }
  &-container {
    width: 100%;
  }
  &-label {
    margin-right: var(--s-basic-spacing);
    word-break: keep-all;
    text-transform: uppercase;
  }
  &-content {
    display: flex;
    justify-content: flex-end;
    align-items: baseline;
    flex-wrap: wrap;
    flex-grow: 1;
    word-break: break-all;
    text-align: right;
  }
  &-value {
    margin-left: auto;
    text-align: right;
    font-weight: 600;
    &-prefix {
      margin-left: auto;
      + .info-line-value {
        margin-left: 0;
      }
    }
  }
  .asset-symbol {
    word-break: keep-all;
    white-space: nowrap;
  }
  .el-tooltip {
    margin-top: -1px;
    margin-right: var(--s-basic-spacing);
    flex-shrink: 0;
    i {
      margin-top: auto;
      margin-bottom: auto;
      display: block;
      color: var(--s-color-base-content-tertiary);
    }
  }
  &-icon {
    position: relative;
    height: var(--s-size-mini);
    width: var(--s-size-mini);
    border-radius: var(--s-border-radius-small);
    color: inherit;
    &:hover {
      background-color: var(--s-color-base-background-hover);
      cursor: pointer;
    }
    &:before {
      position: absolute;
      display: block;
      height: var(--s-icon-font-size-mini);
      width: var(--s-icon-font-size-mini);
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      margin: auto;
      font-size: var(--s-icon-font-size-mini);
    }
    &--left {
      order: -1;
    }
  }
  .el-button {
    margin-left: var(--s-basic-spacing);
    margin-right: 0;
  }
  .token-logo {
    margin-right: var(--s-basic-spacing);
  }
  p {
    font-size: inherit;
  }
  .formatted-amount--fiat-value {
    line-height: inherit;
  }
}
</style>
