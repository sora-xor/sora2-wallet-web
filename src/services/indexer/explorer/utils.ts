import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
