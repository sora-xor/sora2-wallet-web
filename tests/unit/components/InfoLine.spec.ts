import { shallowMount } from '@vue/test-utils';

import { useDescribe, localVue } from '../../utils';
import { MOCK_INFO_LINE } from '../../utils/InfoLineMock';

import InfoLine from '@/components/InfoLine.vue';

useDescribe('InfoLine.vue', InfoLine, () => {
  MOCK_INFO_LINE.map((item) =>
    it(`[${item.title}]: should be rendered correctly`, () => {
      const propsData = {
        label: item.label,
        labelTooltip: item.labelTooltip,
        value: item.value,
        assetSymbol: item.assetSymbol,
        isFormatted: item.isFormatted,
        fiatValue: item.fiatValue,
        valueCanBeHidden: item.valueCanBeHidden,
      };
      const wrapper = shallowMount(InfoLine, {
        localVue,
        propsData,
      });
      expect(wrapper.element).toMatchSnapshot();
    })
  );
});
