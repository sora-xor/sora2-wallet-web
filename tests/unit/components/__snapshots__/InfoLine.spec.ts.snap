// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoLine.vue [Label and Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
     
    <span
      content=""
    >
      <span
        class="info-line-value"
      >
        
          1,234,567.890123
          
        <!---->
      </span>
    </span>
     
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With "-Infinity" Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With "0" Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
     
    <span
      content=""
    >
      <span
        class="info-line-value"
      >
        
          0
          
        <!---->
      </span>
    </span>
     
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With "Infinity" Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With "NaN" Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With Asset Symbol]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
     
    <span
      content=""
    >
      <span
        class="info-line-value"
      >
        
          1,234,567.890123
          
        <span
          class="asset-symbol"
        >
           XOR
        </span>
      </span>
    </span>
     
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With Empty Label and Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With Fiat Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
     
    <span
      content=""
    >
      <span
        class="info-line-value"
      >
        
          1,234,567.890123
          
        <!---->
      </span>
    </span>
     
    <formatted-amount-stub
      assetsymbol=""
      fontweightrate="normal"
      isfiatvalue="true"
      value="1,234,567.890123"
      withleftshift="true"
    />
     
  </div>
</div>
`;

exports[`InfoLine.vue [With Formatted & Hidden value; hide balances button was clicked]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
     
    <formatted-amount-stub
      assetsymbol=""
      class="info-line-value"
      fontsizerate="medium"
      fontweightrate="small"
      value="1,234,567.890123456789"
      valuecanbehidden="true"
    />
     
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With Formatted Value]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
     
    <formatted-amount-stub
      assetsymbol=""
      class="info-line-value"
      fontsizerate="medium"
      fontweightrate="small"
      value="1,234,567.890123456789"
    />
     
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With Hidden value; hide balances button was clicked]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <!---->
   
  <div
    class="info-line-content"
  >
     
    <span
      class="info-line-value"
    >
      ******
    </span>
     
    <!---->
     
  </div>
</div>
`;

exports[`InfoLine.vue [With Label Tooltip]: should be rendered correctly 1`] = `
<div
  class="info-line"
>
   
  <span
    class="info-line-label"
  >
    Label
  </span>
   
  <s-tooltip-stub
    animation="el-fade-in-linear"
    borderradius="mini"
    closedelay="0"
    content="Label tooltip"
    hideafter="0"
    offset="0"
    opendelay="500"
    placement="right-start"
    popperclass="info-tooltip info-tooltip--info-line"
    showarrow="true"
    tabindex="0"
    theme="auto"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="info-16"
      size="14px"
    />
  </s-tooltip-stub>
   
  <div
    class="info-line-content"
  >
     
    <span
      content=""
    >
      <span
        class="info-line-value"
      >
        
          1,234,567.890123
          
        <!---->
      </span>
    </span>
     
    <!---->
     
  </div>
</div>
`;
